### DISCUSSION TITLE
How do I enumerate parameters and other things like that?

### DISCUSSION THREAD

**Discussion by ethindp** [free-audio/clap#425] [Q&A]
So I'm working on writing a small test app to learn the clap standard and how to use it from other things and I have some questions.

Say I want to use something like the Airwindows consolidated plugin as a test plugin (since it's the only clap plugin I have installed that I know about :)). Here's what I've gotten done so far:

1. Load the plugin and find `clap_entry`
2. Call `entry.init(path)`.
3. Call `get_factory` with `CLAP_PLUGIN_FACTORY_ID`.
4. Iterate through all plugin descriptors.

Here's where my questions come in:

1. From what I understand, it is possible to have multiple clap plugins within a single `.clap` file. I believe Airwindows is an example of this, but according to `fac->get_plugin_count` and `fac->get_plugin_descriptor`, only one plugin is present. How does one access "sub-plug-ins" of a clap plug-in? Or am I just mistaken about this particular plug-in?
2. How do we enumerate all the possible parameters of a plug-in? There doesn't seem to be a direct way of doing this from what I can tell, but I very well could be wrong. Do I need to use the preset discovery header to find all presets in the plug-in? Or is there some other function I missed when scanning `plugin.h`?

I'm sure I'll have more questions as I experiment (since I'm planning to use CLAP as a plug-in to a game engine, so I might run into issues there) but from what I can tell, CLAP is so much simpler than VST ever was and I really like it so far!

**Comment by baconpaul** [comment]
Hi!
Airwin con is just one plugin in the clap
The activity you want to undertake is well shown by the clap-info project! Maybe start there?

**Comment by ethindp** [comment]
@baconpaul I figured out params, thanks! So the airwin con is supposed to jsut appear as one plugin? I found a few weird parameters, but it does seem to work:

```
Clap loaded. Clap version: 1.2.0
Found 1 plugins
Plugin com.airwindows.consolidated: Airwindows Consolidated by Airwindows (v. 1.2024.295): Airwindows Consolidated
Plugin requests host ext.: clap.log
Plugin requests host ext.: clap.thread-check
Plugin requests host ext.: clap.thread-pool
Plugin requests host ext.: clap.audio-ports
Plugin requests host ext.: clap.audio-ports-config
Plugin requests host ext.: clap.note-ports
Plugin requests host ext.: clap.timer-support
Plugin requests host ext.: clap.posix-fd-support
Plugin requests host ext.: clap.resource-directory/1
Plugin requests host ext.: clap.latency
Plugin requests host ext.: clap.gui
Plugin requests host ext.: clap.params
Plugin requests host ext.: clap.track-info/1
Plugin requests host ext.: clap.track-info.draft/1
Plugin requests host ext.: clap.state
Plugin requests host ext.: clap.note-name
Plugin requests host ext.: clap.remote-controls/2
Plugin requests host ext.: clap.remote-controls.draft/2
Plugin requests host ext.: clap.voice-info
Plugin requests host ext.: clap.context-menu/1
Plugin requests host ext.: clap.context-menu.draft/0
Plugin requests host ext.: clap.preset-load/2
Plugin requests host ext.: clap.preset-load/2
Plugin initialized
Found 13 parameters
Found param with ID 2944917340: /Replace, min. 0, max. 1, def. 0.5
Found param with ID 2944917341: /Brightness, min. 0, max. 1, def. 0.5
Found param with ID 2944917342: /Detune, min. 0, max. 1, def. 0.5
Found param with ID 2944917343: /Bigness, min. 0, max. 1, def. 1
Found param with ID 2944917344: /Dry/Wet, min. 0, max. 1, def. 1
Found param with ID 2944917345: /-, min. 0, max. 1, def. 0
Found param with ID 2944917346: /-, min. 0, max. 1, def. 0
Found param with ID 2944917347: /-, min. 0, max. 1, def. 0
Found param with ID 2944917348: /-, min. 0, max. 1, def. 0
Found param with ID 2944917349: /-, min. 0, max. 1, def. 0
Found param with ID 2920836328: /Bypass, min. 0, max. 1, def. 0
Found param with ID 100353752: /Input Level, min. 0, max. 1, def. 0.5011872053146362
Found param with ID 3289450511: /Output Level, min. 0, max. 1, def. 0.5011872053146362
```

Mind you, this was just a quick app I through together so it's not meant to be fully like the `clap-info` tool. :) If this is the output I should be seeing then that makes me very happy (the `/`, at least the first, is `module/param_name`). Thank you so much -- CLAP looks like an incredible standard!

**Reply by baconpaul** [reply]
Those don't look weird at all.

Airwin Consolidated has 10 fixed parameters and when you change the effect type, they rename themselves and reset default. The default effect (Galactic) only uses 5 so the other ones get the name '-'

**Reply by baconpaul** [reply]
So yes it looks like you have it right. And glad you are enjoying clap!

between `clap-info` and `clap-wrapper` here, most of the reusable code to write a host is on display (as well as the actual `clap-host` here which is a host)

**Comment by ethindp** [comment]
Oh, thank you, glad to know I'm doing it right! Can't wait to see where this goes -- I'm definitely enjoying every step of the process! :)

**Comment by ethindp** [comment]
Though how do I change the effect type? Is that through the preset-load extension?

**Reply by baconpaul** [reply]
You use the ui.

I had a ui free version with all 420 plugins inside it once (and it wouldn’t be that hard to reinstate) but right now plugin type is not an automatable parameter

**Comment by ethindp** [comment]
Oh okay, it would be nice to reinstate it to have the ability to select the effect type in an automated way, but it's up to you.

**Reply by baconpaul** [reply]
So the reason I didn't do it is because JUCE normalizes parameters to 0...1 (and this plugin uses JUCE) so as Chris adds effects the normalization wouldn't be stable. I tried turning on presets and that fails in the auv2 for a couple of reasons. So basically "the easy ways didn't work" and "I didn't see the need to do a hard thing"

That said, the CLAP state stream is pretty easy to construct if you want.

**Comment by ethindp** [comment]
Oh okay, very interesting, I didn't know that normalization was that unstable in DSP. It would be nice if a version with an effect type parameter was available but your reasoning makes sense and I probably would've made the same decision (because even changing them to percentages and defining some kind of range probably would've just been more annoying than anything else).

**Reply by baconpaul** [reply]
If we were a clap first plugin we could have made a stepped parameter, and similarly if we hand coded a vst. But the juce layer can normalize in some cases which makes automation unstable under min max changes indeed

**Comment by ethindp** [comment]
So just out of curiosity, is there a way to build all 420 plugins into individual CLAPs? Or is only the consolidated one CLAP-capable? (I know, I got lots of questions unrelated to this... Would ask on the discord but I'm not sure which one that is.)

**Reply by baconpaul** [reply]
Surge discord has an air windows channel
It would be easy to make a single clap with 400 plugins yeah. Just typing. It used to work that way but it’s kinda obnoxious in the daws to have that many plugins 
Making 400 distinct clap files is a bit harder but also possible,

**Comment by ethindp** [comment]
Is that the surge synth team discord?

As a suggestion, the DAW version could be as it is but you could always offer different bundles, like one with 400+ plugins or one with each as their own plugin. Though both could be a maintenance disaster I imagine.

**Reply by baconpaul** [reply]
So the single clap multi plugin is barely a problem 

The 405 plugin version is a pain in the neck. The pipeline assembly time would be super duper high and it would be a different “shape” than our current approach (which puts all the fx in a single static lib)
